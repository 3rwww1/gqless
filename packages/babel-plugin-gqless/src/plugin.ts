import jsxSyntax from '@babel/plugin-syntax-jsx'
import { types as t, PluginObj, BabelFileResult, NodePath } from '@babel/core'
import { reactTransform } from './reactTransform'
import { gqlessTransform } from './gqlessTransform'
import { Cache, FileAnalysis } from './analysis'
import { scanImports, ImportCallback } from './scan'

export type State = {
  cwd: string
  file: BabelFileResult
  filename: string

  analysis: FileAnalysis
}

export const plugin = (api: typeof import('@babel/core')): PluginObj<State> => {
  const cache = new Cache(api)

  return {
    name: 'gqless',
    inherits: jsxSyntax,
    pre() {
      const analysis = cache.getPath(this.filename)
      analysis.ast = this.file.ast!
      this.analysis = analysis
    },
    visitor: {
      Program(path, state) {
        scanImports(path, (source): ImportCallback | void => {
          switch (source) {
            case 'gqless':
              return (importName, path) => {
                gqlessTransform(state.analysis, importName, path)
              }

            case '@gqless/react':
              return (importName, path) => {
                reactTransform(importName, path)
              }
          }
        })
      },
    },
  }
}
