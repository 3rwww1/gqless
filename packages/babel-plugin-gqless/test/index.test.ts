import path from 'path'
// @ts-ignore
import pluginTester from 'babel-plugin-tester'
import plugin from 'babel-plugin-gqless'

const projectRoot = path.join(__dirname, '../../')

expect.addSnapshotSerializer({
  print(val) {
    return val.split(projectRoot).join('<PROJECT_ROOT>/')
  },
  test(val) {
    return typeof val === 'string'
  },
})

const fixture = (filename: string) => ({
  fixture: require.resolve(`./fixtures/${filename}`),
})

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename },
  tests: {
    'supports preload': fixture('preload'),
    'inserts name option for components': fixture('componentName'),
    'inserts name for useVariable': fixture('variableName'),
    'inserts name for useFragment': fixture('fragmentName'),
  },
})
