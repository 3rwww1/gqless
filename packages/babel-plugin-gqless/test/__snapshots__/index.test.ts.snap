// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gqless test: test 1`] = `

import { graphql } from '@gqless/react'

export const Component = graphql(() => {
  return <div>test</div>
})

      ↓ ↓ ↓ ↓ ↓ ↓

import { graphql } from '@gqless/react';
export const Component = graphql(() => {
  return <div>test</div>;
});

`;
